### PR TITLE
`PILOT_FILTER_GATEWAY_CLUSTER_CONFIG`: fix false negatives for AUTO_PASSTHROUGH

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
Auto-passthrough gateways do not require virtual services, we should push when related services are changed

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -800,15 +800,27 @@ func (ps *PushContext) GatewayServices(proxy *Proxy) []*Service {
...
+					if hostname.SubsetOf(host.Name(h)) {
+						return true
+					}
+				}
+			}
+		}
+	}
+	for _, g := range gw.GatewayNameForServer {
+		if hosts := ps.virtualServiceIndex.destinationsByGateway[g]; hosts != nil {
+			if hosts.Contains(string(hostname)) {
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:823]: I know this is existing code, but how do we handle a wildcard VS?

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (dddddai) [reply pilot/pkg/model/push_context.go:823]: Sorry I don't get it, could you give an example?

**Comment by hzxuzhonghu** [comment]
If this is fixing a bug, should backport

File: pilot/pkg/model/push_context.go [hunk 2]
```diff
@@ -800,15 +800,27 @@ func (ps *PushContext) GatewayServices(proxy *Proxy) []*Service {
...
 	// MergedGateway will be nil when there are no configs in the
 	// system during initial installation.
-	if proxy.MergedGateway == nil {
+	if gw == nil {
 		return false
 	}
-	for _, gw := range proxy.MergedGateway.GatewayNameForServer {
-		if hosts := ps.virtualServiceIndex.destinationsByGateway[gw]; hosts != nil {
-			if hosts.Contains(hostname) {
+	if gw.ContainsAutoPassthroughGateways {
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/push_context.go:810]: If IIRC, this has several bugs, can we check ` proxy.SidecarScope.GetService(hostname)`

Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:810]: Seems donot need to check `ServiceAttachedToGateway`

Comment (dddddai) [reply pilot/pkg/model/push_context.go:810]: > If IIRC, this has several bugs, can we check ` proxy.SidecarScope.GetService(hostname)`

Already checked [here](https://github.com/istio/istio/pull/42721/files/3d8a2d0b57b1f7f7903d4cf514166c41234090cb#diff-eea91d8b0993477d5d708587b9f186b1efbc625019e9c6b21c1217bd4d4b8703R78)

> Seems donot need to check `ServiceAttachedToGateway`

Yes we could just return true for auto-passthrough gateways, filtering hosts gives a more precise check though

Comment (dddddai) [reply pilot/pkg/model/push_context.go:810]: I think we are doing the same check when building filters?
https://github.com/istio/istio/blob/1.16.1/pilot/pkg/networking/core/v1alpha3/gateway.go#L885-L886

Comment (dddddai) [reply pilot/pkg/model/push_context.go:810]: The only bug case I can think of is envoy filters, let's just return true anyways

Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:810]: This is known one, the potential risk is unknown. SO suggest making it simple

**Pull request merged**